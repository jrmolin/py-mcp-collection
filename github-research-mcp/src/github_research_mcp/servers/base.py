from typing import TYPE_CHECKING, Any

from fastmcp.server.dependencies import get_context
from fastmcp.utilities.logging import get_logger
from githubkit.github import GitHub
from mcp.types import ContentBlock, SamplingMessage, TextContent
from pydantic import BaseModel

from github_research_mcp.models.graphql.queries import BaseGqlQuery
from github_research_mcp.servers.shared.utility import estimate_model_tokens

if TYPE_CHECKING:
    from fastmcp.server import Context

logger = get_logger(__name__)


class BaseResponseModel(BaseModel):
    def estimate_tokens(self) -> int:
        return estimate_model_tokens(self)


class BaseServer:
    github_client: GitHub[Any]

    async def _sample(self, system_prompt: str, messages: str | list[str], max_tokens: int = 2000, temperature: float = 0.0) -> str:
        """Sample a response from the server."""

        context: Context = get_context()

        if isinstance(messages, str):
            messages = [messages]

        sampling_messages: str | list[str | SamplingMessage] = [
            SamplingMessage(role="user", content=TextContent(type="text", text=message)) for message in messages
        ]

        summary: ContentBlock = await context.sample(
            system_prompt=system_prompt,
            messages=sampling_messages,
            temperature=temperature,
            max_tokens=max_tokens,
        )

        if not isinstance(summary, TextContent):
            msg = "The sampling call failed to generate a valid text summary of the issue."
            raise TypeError(msg)

        return summary.text

    async def _perform_graphql_query[T: BaseGqlQuery](self, query_model: type[T], variables: dict[str, Any]) -> T:
        """Perform a GraphQL query and return the response as a model."""

        logger.info(f"Executing GraphQL query {query_model.__name__} with variables {variables}")

        raw_response = await self.github_client.async_graphql(
            query=query_model.graphql_query(),
            variables=variables,
        )

        response_model = query_model.model_validate(raw_response)
        response_size = estimate_model_tokens(response_model)

        logger.info(f"Completed GraphQL query {query_model.__name__} for with variables {variables} returned {response_size} tokens.")

        return response_model
