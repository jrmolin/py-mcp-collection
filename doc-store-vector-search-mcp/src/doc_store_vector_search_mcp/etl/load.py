from abc import ABC, abstractmethod
from collections.abc import AsyncGenerator
from typing import Any

from langchain_community.document_loaders import DirectoryLoader as LangchainDirectoryLoader
from langchain_community.document_loaders import RecursiveUrlLoader as LangchainRecursiveUrlLoader
from langchain_community.document_loaders import WebBaseLoader as LangchainWebBaseLoader
from langchain_community.document_loaders.json_loader import JSONLoader as LangchainJSONLoader
from langchain_community.document_loaders.sitemap import SitemapLoader as LangchainSitemapLoader
from langchain_core.documents import Document


type LoaderTypes = WebPageLoader | SiteMapLoader | RecursiveWebLoader | DirectoryLoader | JSONJQLoader


class WebPageLoader:

    @classmethod
    async def load(cls, url: str, **kwargs: Any) -> AsyncGenerator[list[Document], None]:
        loader = LangchainWebBaseLoader(url, **kwargs)

        async for document in loader.alazy_load():
            yield document


class SiteMapLoader:

    @classmethod
    async def load(cls, url: str, filter_urls: list[str] | None = None, **kwargs: Any) -> AsyncGenerator[list[Document], None]:
        loader = LangchainSitemapLoader(url, filter_urls=filter_urls, **kwargs)

        async for document in loader.alazy_load():
            yield document


class RecursiveWebLoader:

    async def load(cls, url: str, **kwargs: Any) -> AsyncGenerator[list[Document], None]:
        loader = LangchainRecursiveUrlLoader(url, use_async=True, **kwargs)

        async for document in loader.alazy_load():
            yield document


class DirectoryLoader:

    @classmethod
    async def load(
        cls, directory_path: str, glob: str = "**/*.{md,mdx,txt}", **kwargs: Any
    ) -> AsyncGenerator[list[Document], None]:
        text_loader_kwargs = {"autodetect_encoding": True}
        loader = LangchainDirectoryLoader(
            directory_path, glob=glob, silent_errors=False, loader_kwargs=text_loader_kwargs, **kwargs
        )

        async for document in loader.alazy_load():
            yield document


class JSONJQLoader:

    @classmethod
    def load(cls, json_path: str, jq_schema: str = ".items[]", content_key: str = "text", **kwargs: Any) -> AsyncGenerator[list[Document], None]:
        loader = LangchainJSONLoader(json_path, jq_schema=jq_schema, content_key=content_key, **kwargs)

        yield from loader.lazy_load()
