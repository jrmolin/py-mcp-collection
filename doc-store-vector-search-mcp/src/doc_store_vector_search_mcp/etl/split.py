from abc import ABC, abstractmethod
from typing import Any, TypeAlias

import langchain_text_splitters
from langchain_core.documents import Document
from langchain_text_splitters import (
    MarkdownHeaderTextSplitter,
    RecursiveCharacterTextSplitter,
    RecursiveJsonSplitter,
)
from typing_extensions import TypeVar

T = TypeVar("T")


class Splitter(ABC):
    @abstractmethod
    def split_to_documents(self, to_split: T) -> list[Document]:
        pass


CodeSplitterLanguage: TypeAlias = langchain_text_splitters.Language  # noqa: UP040


def tag_document_order(documents: list[Document]) -> list[Document]:
    for i, document in enumerate(documents):
        document.metadata["order"] = i
    return documents


class JsonSplitter(Splitter):
    def __init__(self):
        self.splitter = RecursiveJsonSplitter(
            max_chunk_size=500,
            min_chunk_size=25,
        )

    def split_to_documents(self, to_split: dict[str, Any]) -> list[Document]:
        return tag_document_order(self.splitter.create_documents([to_split], convert_lists=True))


class MarkdownSplitter:
    def __init__(self):
        self.header_splitter = MarkdownHeaderTextSplitter(
            headers_to_split_on=[
                ("#", "Header 1"),
                ("##", "Header 2"),
            ],
            strip_headers=False,
        )
        self.chunk_splitter = RecursiveCharacterTextSplitter(
            chunk_size=300,
            chunk_overlap=25,
        )

    def split_to_documents(self, text: str) -> list[Document]:
        per_header_documents: list[Document] = self.header_splitter.split_text(text)
        chunked_documents: list[Document] = []

        for document in per_header_documents:
            chunked_documents.extend(self.chunk_splitter.split_documents([document]))

        return tag_document_order(chunked_documents)


class CodeSplitter:
    def __init__(self, language: CodeSplitterLanguage):
        self.splitter = RecursiveCharacterTextSplitter.from_language(
            language,
            chunk_size=300,
            chunk_overlap=25,
        )

    def split_to_documents(self, text: str) -> list[Document]:
        chunks = self.splitter.split_text(text)
        return tag_document_order(self.splitter.create_documents(chunks))
