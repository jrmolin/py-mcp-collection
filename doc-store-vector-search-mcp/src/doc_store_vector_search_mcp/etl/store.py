from typing import Any, TypeVar

import duckdb
import torch
from langchain_community.vectorstores import DuckDB, ElasticsearchStore
from langchain_core.documents import Document
from langchain_core.embeddings import Embeddings
from pydantic import BaseModel
from sentence_transformers import SentenceTransformer
from transformers import AutoModel, AutoTokenizer

type VectorStoreTypes = DuckDB | ElasticsearchStore


class ElasticsearchStoreSettings(BaseModel):
    es_url: str
    es_index: str

    def connect(self) -> None:
        pass


class DuckDBSettings(BaseModel):
    db_path: str

    def connect(self) -> duckdb.DuckDBPyConnection:
        return duckdb.connect(database=self.db_path, config={"enable_external_access": "false"})


type VectorStoreSettings = ElasticsearchStoreSettings | DuckDBSettings

T = TypeVar("T", bound=VectorStoreTypes)
S = TypeVar("S", bound=VectorStoreSettings)


class TextEmbedding(Embeddings):
    def __init__(self, model: str = "all-MiniLM-L6-v2"):
        self.model = SentenceTransformer(model)

    def embed(self, text: str) -> list[float]:
        return self.model.encode([text])[0].tolist()

    def embed_documents(self, texts: list[str]) -> list[list[float]]:
        return self.model.encode(texts).tolist()

    def embed_query(self, text: str) -> list[float]:
        return self.model.encode([text])[0].tolist()


class CodeEmbedding(Embeddings):
    def __init__(self):
        self.device = "cpu"

        self.tokenizer = AutoTokenizer.from_pretrained("Salesforce/codet5p-110m-embedding", trust_remote_code=True)
        self.model = AutoModel.from_pretrained("Salesforce/codet5p-110m-embedding", trust_remote_code=True).to(self.device)

        self.torch = torch

    def embed(self, text: str) -> list[float]:
        inputs = self.tokenizer.encode(text, return_tensors="pt").to(self.device)
        with self.torch.no_grad():
            embedding = self.model(inputs)[0]
        return embedding.cpu().numpy().tolist()

    def embed_documents(self, texts: list[str]) -> list[list[float]]:
        return [self.embed(text) for text in texts]

    def embed_query(self, text: str) -> list[float]:
        return self.embed(text)


class VectorStoreManager[T: VectorStoreTypes, S: VectorStoreSettings]:
    def __init__(self, project_name: str, kb_id: str, code_settings: S, document_settings: S, vector_store_class: type[T]):
        self.project_name = project_name
        self.kb_id = kb_id

        self.code_embedding = CodeEmbedding()
        self.document_embedding = TextEmbedding()

        self.code_vector_store: T = vector_store_class(embedding=self.code_embedding, connection=code_settings.connect())  # type: ignore
        self.document_vector_store: T = vector_store_class(embedding=self.document_embedding, connection=document_settings.connect())  # type: ignore

    async def add_code_documents(self, documents: list[Document], metadata: dict[str, Any]):
        [document.metadata.update(metadata or {}) for document in documents]

        await self.code_vector_store.aadd_documents(documents)

    async def add_documents(self, documents: list[Document], metadata: dict[str, Any]):
        [document.metadata.update(metadata or {}) for document in documents]
        await self.document_vector_store.aadd_documents(documents)


class ProjectVectorStoreManager[T: VectorStoreTypes, S: VectorStoreSettings]:
    def __init__(self, project_name: str, code_vector_store: VectorStoreManager[T, S], document_vector_store: VectorStoreManager[T, S]):
        self.project_name = project_name
        self.code_vector_store = code_vector_store
        self.document_vector_store = document_vector_store

    async def add_code_documents(self, documents: list[Document], metadata: dict[str, Any]):
        updated_metadata = {
            **metadata,
            "project_name": self.project_name,
        }
        await self.code_vector_store.add_code_documents(documents, updated_metadata)

    async def add_markdown_documents(self, documents: list[Document], metadata: dict[str, Any]):
        updated_metadata = {
            **metadata,
            "project_name": self.project_name,
        }

        await self.document_vector_store.add_documents(documents, updated_metadata)


class KnowledgeBaseVectorStoreManager:
    def __init__(self, kb_id: str, project_vector_store: ProjectVectorStoreManager):
        self.kb_id = kb_id
        self.project_vector_store = project_vector_store

    async def add_code_documents(self, documents: list[Document], metadata: dict[str, Any]):
        updated_metadata = {
            **metadata,
            "kb_id": self.kb_id,
        }
        await self.project_vector_store.add_code_documents(documents, updated_metadata)

    async def add_markdown_documents(self, documents: list[Document], metadata: dict[str, Any]):
        updated_metadata = {
            **metadata,
            "kb_id": self.kb_id,
        }

        await self.project_vector_store.add_markdown_documents(documents, updated_metadata)

    @classmethod
    def from_project_vector_store(cls, kb_id: str, project_vector_store: ProjectVectorStoreManager) -> "KnowledgeBaseVectorStoreManager":
        return cls(
            kb_id=kb_id,
            project_vector_store=project_vector_store,
        )
