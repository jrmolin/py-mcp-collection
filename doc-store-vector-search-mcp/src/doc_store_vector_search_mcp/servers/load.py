from collections.abc import AsyncGenerator
from datetime import UTC, datetime

from fastmcp.contrib.mcp_mixin.mcp_mixin import MCPMixin, mcp_tool
from html_to_markdown import markdownify
from langchain_core.documents import Document
from pydantic import BaseModel, Field, HttpUrl, field_serializer

from doc_store_vector_search_mcp.etl.load import DirectoryLoader, RecursiveWebLoader, WebPageLoader
from doc_store_vector_search_mcp.etl.split import HtmlSplitter, MarkdownSplitter, SemanticSplitter
from doc_store_vector_search_mcp.etl.store import KnowledgeBaseVectorStoreManager, ProjectVectorStoreManager
from doc_store_vector_search_mcp.logging.util import BASE_LOGGER

logger = BASE_LOGGER.getChild("load")


class StartLoad(BaseModel):
    start_time: datetime = Field(default_factory=lambda: datetime.now(UTC))


class LoadSummary(BaseModel):
    source_documents: int
    documents: int
    took: float

    @classmethod
    def from_start_load(cls, start_load: StartLoad, source_documents: int, documents: int) -> "LoadSummary":
        return cls(
            source_documents=source_documents,
            documents=documents,
            took=(datetime.now(UTC) - start_load.start_time).total_seconds(),
        )


class DocumentKnowledgeBaseMetadata(BaseModel):
    knowledge_base: str
    project: str


class DocumentMetadata(DocumentKnowledgeBaseMetadata):
    target: str
    fetched: datetime = Field(default_factory=lambda: datetime.now(UTC))

    @field_serializer("fetched")
    def serialize_fetched(self, fetched: datetime, _info):
        return fetched.timestamp()

    @classmethod
    def from_document(cls, kb_metadata: DocumentKnowledgeBaseMetadata, target: str, fetched: datetime) -> "DocumentMetadata":
        return cls(knowledge_base=kb_metadata.knowledge_base, project=kb_metadata.project, target=target, fetched=fetched)


class DocumentServer(MCPMixin):
    def __init__(self, project_name: str, project_vectorstore: ProjectVectorStoreManager):
        self.project_name = project_name
        self.project_vectorstore = project_vectorstore
        self.underlying_vector_store = project_vectorstore.document_vector_store

    def _get_started(self, knowledge_base: str) -> tuple[KnowledgeBaseVectorStoreManager, StartLoad, DocumentKnowledgeBaseMetadata]:
        knowledge_base_vectorstore = KnowledgeBaseVectorStoreManager.from_project_vector_store(knowledge_base, self.project_vectorstore)
        start_load = StartLoad()
        kb_metadata = DocumentKnowledgeBaseMetadata(knowledge_base=knowledge_base, project=self.project_name)

        logger.info(f"Starting load for knowledge base {knowledge_base} in project {self.project_name}")

        return knowledge_base_vectorstore, start_load, kb_metadata

    @classmethod
    async def load_directory(
        cls, directory: str
    ) -> AsyncGenerator[
        Document,
        None,
    ]:
        async for document in DirectoryLoader.load(directory_path=directory, glob=["**/*.md", "**/*.txt", "**/*.mdx"]):
            yield document

    @mcp_tool()
    async def load_directories(self, knowledge_base: str, directories: list[str]) -> LoadSummary:
        """Loads all documents in the given directories into the knowledge base.

        Args:
            knowledge_base: The name of theknowledge base to load the documents into.
            directories: The directories to load the documents from.

        Returns:
            A summary of the load including the number of source documents, documents, and chunks.
        """

        logger.info(f"Loading directories {directories} for knowledge base {knowledge_base} in project {self.project_name}")

        knowledge_base_vectorstore, start_load, kb_metadata = self._get_started(knowledge_base)

        source_documents = 0
        document_count = 0
        chunk_count = 0

        splitter = MarkdownSplitter()

        for directory in directories:
            async for document in self.load_directory(directory):
                source_documents += 1

                document_metadata = DocumentMetadata.from_document(kb_metadata, directory, datetime.now(UTC))
                split_documents = splitter.split_to_documents(document.page_content)

                document_count += 1
                chunk_count += len(split_documents)

                logger.info(f"Adding {len(split_documents)} chunks to knowledge base {knowledge_base} in project {self.project_name}")
                await knowledge_base_vectorstore.add_markdown_documents(
                    split_documents,
                    {
                        **document_metadata.model_dump(),
                        **document.metadata,
                    },
                )

        load_summary = LoadSummary.from_start_load(start_load, source_documents, document_count)
        logger.info(f"Load summary: {load_summary}")

        return load_summary

    async def process_web_document(self, html_splitter: HtmlSplitter, document: Document) -> list[Document]:
        split_lines = document.page_content.splitlines()
        split_lines = [line for line in split_lines if line.strip() != ""]
        # remove consecutive duplicate lines
        split_lines = [line for i, line in enumerate(split_lines) if i == 0 or line != split_lines[i - 1]]

        new_documents = html_splitter.split_to_documents("\n".join(split_lines))
        [new_document.metadata.update(document.metadata) for new_document in new_documents]
        return new_documents

    async def process_web_document_as_markdown(self, markdown_splitter: MarkdownSplitter, document: Document) -> list[Document]:
        markdown = markdownify(document.page_content)
        print(f"Markdown: {markdown}")
        new_documents = markdown_splitter.split_to_documents(markdown)
        [print(f"New document: {new_document.page_content}") for new_document in new_documents]
        [new_document.metadata.update(document.metadata) for new_document in new_documents]
        return new_documents

    @mcp_tool()
    async def load_webpage(self, knowledge_base: str, urls: list[HttpUrl], recursive: bool = False) -> LoadSummary:
        """Loads the given URLs into the knowledge base.

        Args:
            knowledge_base: The name of the knowledge base to load the documents into.
            urls: The URLs to load the documents from.
            recursive: Whether to follow links on the page and load them recursively.

        Returns:
            A summary of the load including the number of source documents, documents, and chunks.
        """

        logger.info(f"Loading {len(urls)} URLs for knowledge base {knowledge_base} in project {self.project_name}")

        knowledge_base_vectorstore, start_load, kb_metadata = self._get_started(knowledge_base)

        source_documents = 0
        document_count = 0

        html_splitter = HtmlSplitter()
        semantic_splitter = SemanticSplitter(self.underlying_vector_store.document_semantic_splitter.embeddings)

        loader = WebPageLoader if not recursive else RecursiveWebLoader

        for url in urls:
            async for raw_html_document in loader.load(url=str(url)):
                source_documents += 1
                document_metadata = DocumentMetadata.from_document(kb_metadata, str(url), datetime.now(UTC))

                html_documents = html_splitter.split_documents([raw_html_document])

                semantically_grouped_html_documents = semantic_splitter.split_documents(html_documents)

                prepared_documents = [
                    self.underlying_vector_store.prepare_document_for_embedding(document)
                    for document in semantically_grouped_html_documents
                ]

                document_count += len(prepared_documents)

                if len(prepared_documents) == 0:
                    logger.warning(f"No documents to add for {url}")
                    continue

                logger.info(f"Adding {len(prepared_documents)} chunks to knowledge base {knowledge_base} in project {self.project_name}")
                await knowledge_base_vectorstore.add_markdown_documents(
                    prepared_documents,
                    {
                        **document_metadata.model_dump(),
                        **raw_html_document.metadata,
                    },
                )

        return LoadSummary.from_start_load(start_load, source_documents, document_count)
