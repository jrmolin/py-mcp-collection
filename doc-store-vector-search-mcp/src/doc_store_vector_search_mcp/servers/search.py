from fastmcp.contrib.mcp_mixin.mcp_mixin import MCPMixin, mcp_tool
from langchain_core.documents import Document
from pydantic import BaseModel

from doc_store_vector_search_mcp.etl.store import ProjectVectorStoreManager, SearchResult

from doc_store_vector_search_mcp.logging.util import BASE_LOGGER

logger = BASE_LOGGER.getChild("search")


# class SearchResult(BaseModel):
#     content: str
#     title: str
#     url: str
#     score: float

#     @classmethod
#     def from_document(cls, document: Document) -> "SearchResult":
#         return cls(
#             content=document.page_content,
#             title=document.metadata.get("title", "<no title>"),
#             url=document.metadata.get("url", "<no url>"),
#             score=document.metadata.get("_similarity_score", 0),
#         )

class SearchServer(MCPMixin):
    def __init__(self, project_name: str, project_vectorstore: ProjectVectorStoreManager):
        self.project_name = project_name
        self.project_vectorstore = project_vectorstore

    @mcp_tool()
    async def search_documents(self, query: str) -> list[SearchResult]:
        logger.info(f"Searching {self.project_name} documents for query: {query}")
        return await self.project_vectorstore.search_documents(query)

    # @mcp_tool()
    # async def search_code(self, query: str) -> list[SearchResult]:
    #     logger.info(f"Searching {self.project_name} code for query: {query}")
    #     return await self.project_vectorstore.search_code(query)
