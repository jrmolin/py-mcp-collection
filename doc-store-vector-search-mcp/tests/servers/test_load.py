from pathlib import Path
from typing import Any
from unittest.mock import patch

import pytest
from langchain_core.documents import Document
from syrupy.assertion import SnapshotAssertion

from doc_store_vector_search_mcp.etl.store import (
    KnowledgeBaseVectorStoreManager,
)
from doc_store_vector_search_mcp.servers.load import DocumentServer

MEATY_CONTENT_CHUNK = """
This is a chunk of meaty content.

It has multiple lines. The lines are separated by a newline character. And they get pretty long.
This is another line. That is also pretty long. It is also separated by a newline character.

And it has multiple paragraphs.
"""


@pytest.fixture
def test_files(tmp_path: Path):
    md_file = tmp_path / "file.md"
    md_file.write_text(f"""
# Header 1a

{MEATY_CONTENT_CHUNK}

## Header 2

{MEATY_CONTENT_CHUNK}

### Header 3

{MEATY_CONTENT_CHUNK}

#### Header 4

{MEATY_CONTENT_CHUNK}

# Header 1b

{MEATY_CONTENT_CHUNK}

## Header 2b

{MEATY_CONTENT_CHUNK}
""")
    txt_file = tmp_path / "file.txt"
    txt_file.write_text("Just a plain text file.")
    return tmp_path


@pytest.fixture
def dummy_project_vectorstore():
    # This can be None or a simple object, since we patch the only method used
    return object()


@pytest.fixture(autouse=True)
def freezer_time(freezer):
    freezer.move_to("2025-01-01 12:00:00")


@pytest.mark.asyncio
async def test_load_directories_loader_splitter_combo(test_files, dummy_project_vectorstore, snapshot: SnapshotAssertion):
    server = DocumentServer(project_name="test_project", project_vectorstore=dummy_project_vectorstore)

    captured = []

    async def fake_add_markdown_documents(self, docs: list[Document], metadata: dict[str, Any]):  # noqa: ARG001
        metadata["source"] = "test_files"
        for d in docs:
            d.metadata["source"] = "test_files"
        captured.append(
            {
                "docs": [{"page_content": d.page_content, "metadata": d.metadata} for d in docs],
                "metadata": metadata,
            }
        )

    with patch.object(KnowledgeBaseVectorStoreManager, "add_markdown_documents", new=fake_add_markdown_documents):
        summary = await server.load_directories(
            knowledge_base="test_kb",
            directories=[str(test_files)],
        )

    snapshot.assert_match(
        {
            "summary": summary,
            "calls": captured,
        }
    )


async def test_load_webpage(dummy_project_vectorstore, snapshot: SnapshotAssertion):
    server = DocumentServer(project_name="test_project", project_vectorstore=dummy_project_vectorstore)

    captured = []

    async def fake_add_markdown_documents(self, docs: list[Document], metadata: dict[str, Any]):  # noqa: ARG001
        metadata["source"] = "test_files"
        for d in docs:
            d.metadata["source"] = "test_files"
        captured.append(
            {
                "docs": [{"page_content": d.page_content, "metadata": d.metadata} for d in docs],
                "metadata": metadata,
            }
        )

    with patch.object(KnowledgeBaseVectorStoreManager, "add_markdown_documents", new=fake_add_markdown_documents):
        summary = await server.load_webpage(
            knowledge_base="test_kb",
            urls=["https://www.example.com"],
        )

    snapshot.assert_match(
        {
            "summary": summary,
            "calls": captured,
        }
    )


async def test_load_webpage_recursive(dummy_project_vectorstore, snapshot: SnapshotAssertion):
    server = DocumentServer(project_name="test_project", project_vectorstore=dummy_project_vectorstore)

    captured = []

    async def fake_add_markdown_documents(self, docs: list[Document], metadata: dict[str, Any]):  # noqa: ARG001
        metadata["source"] = "test_files"
        for d in docs:
            d.metadata["source"] = "test_files"
        captured.append(
            {
                "docs": [{"page_content": d.page_content, "metadata": d.metadata} for d in docs],
                "metadata": metadata,
            }
        )

    with patch.object(KnowledgeBaseVectorStoreManager, "add_markdown_documents", new=fake_add_markdown_documents):
        summary = await server.load_webpage(
            knowledge_base="test_kb",
            urls=["https://www.example.com"],
            recursive=True,
        )

    snapshot.assert_match(
        {
            "summary": summary,
            "calls": captured,
        }
    )
