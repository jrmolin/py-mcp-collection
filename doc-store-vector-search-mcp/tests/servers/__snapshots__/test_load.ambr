# serializer version: 1
# name: test_load_directories_loader_splitter_combo
  dict({
    'calls': list([
      dict({
        'docs': list([
          dict({
            'metadata': dict({
              'order': 0,
              'source': 'test_files',
            }),
            'page_content': '''
              Header 1a  
              This is a chunk of meaty content.  
              It has multiple lines. The lines are separated by a newline character. And they get pretty long. This is another line. That is also pretty long. It is also separated by a newline character.  
              And it has multiple paragraphs.  
              Header 2
            ''',
          }),
          dict({
            'metadata': dict({
              'order': 1,
              'source': 'test_files',
            }),
            'page_content': '''
              Header 2  
              This is a chunk of meaty content.  
              It has multiple lines. The lines are separated by a newline character. And they get pretty long. This is another line. That is also pretty long. It is also separated by a newline character.  
              And it has multiple paragraphs.  
              Header 3
            ''',
          }),
          dict({
            'metadata': dict({
              'order': 2,
              'source': 'test_files',
            }),
            'page_content': '''
              Header 3  
              This is a chunk of meaty content.  
              It has multiple lines. The lines are separated by a newline character. And they get pretty long. This is another line. That is also pretty long. It is also separated by a newline character.  
              And it has multiple paragraphs.  
              Header 4
            ''',
          }),
          dict({
            'metadata': dict({
              'order': 3,
              'source': 'test_files',
            }),
            'page_content': '''
              Header 4  
              This is a chunk of meaty content.  
              It has multiple lines. The lines are separated by a newline character. And they get pretty long. This is another line. That is also pretty long. It is also separated by a newline character.  
              And it has multiple paragraphs.  
              Header 1b
            ''',
          }),
          dict({
            'metadata': dict({
              'order': 4,
              'source': 'test_files',
            }),
            'page_content': '''
              Header 1b  
              This is a chunk of meaty content.  
              It has multiple lines. The lines are separated by a newline character. And they get pretty long. This is another line. That is also pretty long. It is also separated by a newline character.  
              And it has multiple paragraphs.  
              Header 2b
            ''',
          }),
          dict({
            'metadata': dict({
              'order': 5,
              'source': 'test_files',
            }),
            'page_content': '''
              Header 2b  
              This is a chunk of meaty content.  
              It has multiple lines. The lines are separated by a newline character. And they get pretty long. This is another line. That is also pretty long. It is also separated by a newline character.  
              And it has multiple paragraphs.
            ''',
          }),
        ]),
        'metadata': dict({
          'fetched': FakeDatetime(2025, 1, 1, 12, 0, tzinfo=datetime.timezone.utc),
          'knowledge_base': 'test_kb',
          'project': 'test_project',
          'source': 'test_files',
        }),
      }),
      dict({
        'docs': list([
          dict({
            'metadata': dict({
              'order': 0,
              'source': 'test_files',
            }),
            'page_content': 'Just a plain text file.',
          }),
        ]),
        'metadata': dict({
          'fetched': FakeDatetime(2025, 1, 1, 12, 0, tzinfo=datetime.timezone.utc),
          'knowledge_base': 'test_kb',
          'project': 'test_project',
          'source': 'test_files',
        }),
      }),
    ]),
    'summary': LoadSummary(source_documents=2, documents=2, chunks=7, took=0.0),
  })
# ---
# name: test_load_webpage
  dict({
    'calls': list([
      dict({
        'docs': list([
          dict({
            'metadata': dict({
              'Header 1': '#TITLE#',
              'order': 0,
              'source': 'test_files',
            }),
            'page_content': '''
              Example Domain
              This domain is for use in illustrative examples in documents. You may use this
                  domain in literature without prior coordination or asking for permission.
              More information...
            ''',
          }),
        ]),
        'metadata': dict({
          'fetched': FakeDatetime(2025, 1, 1, 12, 0, tzinfo=datetime.timezone.utc),
          'knowledge_base': 'test_kb',
          'project': 'test_project',
          'source': 'test_files',
        }),
      }),
    ]),
    'summary': LoadSummary(source_documents=1, documents=1, chunks=1, took=0.0),
  })
# ---
# name: test_load_webpage_recursive
  dict({
    'calls': list([
      dict({
        'docs': list([
          dict({
            'metadata': dict({
              'Header 1': 'Example Domain',
              'order': 0,
              'source': 'test_files',
            }),
            'page_content': '''
              Example Domain 
               This domain is for use in illustrative examples in documents. You may use this
                  domain in literature without prior coordination or asking for permission. 
               More information...
            ''',
          }),
        ]),
        'metadata': dict({
          'fetched': FakeDatetime(2025, 1, 1, 12, 0, tzinfo=datetime.timezone.utc),
          'knowledge_base': 'test_kb',
          'project': 'test_project',
          'source': 'test_files',
        }),
      }),
    ]),
    'summary': LoadSummary(source_documents=1, documents=1, chunks=1, took=0.0),
  })
# ---
